agent/cache-types/streaming_health_services.go

```diff
@@ -22,170 +20,65 @@ const (
-		}
-		go sub.run()
-	} else {
-		sub = opts.LastResult.State.(*subscriber)
+	subscribeReq := stream.SubscribeRequest{
+		Topic: stream.Topic_ServiceHealth,
+		Key:   reqReal.ServiceName,
 	}
-
-	result.State = sub
-
-	// If the requested index is lower than what we've already seen, return immediately.
-	sub.lock.RLock()
-	if sub.lastResult.Index > opts.MinIndex {
-		result.Value = &sub.lastResult
-		result.Index = sub.lastResult.Index
-		sub.lock.RUnlock()
-		return result, nil
+	handler := healthServicesHandler{
+		state: make(map[string]structs.CheckServiceNode),
 	}
-	sub.lock.RUnlock()
-
-	// Wait for an update based on the index we're blocking on.
-	timeout := time.After(StreamTimeout)
-WAIT:
-	select {
-	case r := <-sub.resultCh:
-		switch reply := r.(type) {
-		case *structs.IndexedCheckServiceNodes:
-			if reply.Index <= opts.MinIndex {
-				goto WAIT
-			}
-			result.Value = reply
-			result.Index = reply.Index
-		case error:
-			result.Value = reply
+	indexFunc := func(v interface{}) uint64 {
+		if v, ok := v.(*structs.IndexedCheckServiceNodes); ok {
+			return v.Index
 		}
-	case <-timeout:
-		sub.lock.RLock()
-		result.Value = &sub.lastResult
-		result.Index = sub.lastResult.Index
-		sub.lock.RUnlock()
+		return 0
 	}
 
-	return result, nil
+	return watchSubscriber(c.Client, opts, subscribeReq, &handler, indexFunc)
 }
 
 func (c *StreamingHealthServices) SupportsBlocking() bool {
 	return true
 }
 
-// subscriber runs a subscription for a single topic/key combination.
-type subscriber struct {
-	client     stream.ConsulClient
-	service    string
-	lastResult structs.IndexedCheckServiceNodes
-	lastUpdate time.Time
-	resultCh   chan interface{}
-
-	cancelFunc func()
-
-	lock sync.RWMutex
-}
-
-// Close stops the subscriber by cancelling the context.
-func (s *subscriber) Close() error {
-	s.cancelFunc()
-	return nil
+// healthServicesHandler maintains a view of the health of a service
+// based on incoming events from a Subscriber.
+type healthServicesHandler struct {
+	state map[string]structs.CheckServiceNode
 }
 
-// run starts a Subscribe call for the subscriber's key, receives the initial
-// snapshot of state and then sends the state to the watcher as soon as its
-// watched index requirement is met.
-func (s *subscriber) run() {
-	ctx, cancel := context.WithCancel(context.Background())
-	s.cancelFunc = cancel
-	defer s.cancelFunc()
-
-	// Start a new Subscribe call.
-	streamHandle, err := s.client.Subscribe(ctx, &stream.SubscribeRequest{
-		Topic: stream.Topic_ServiceHealth,
-		Key:   s.service,
-	})
-
-	// If something went wrong setting up the stream, return the error.
-	if err != nil {
-		// Do a blocking send since there should always be a Fetch goroutine watching
-		// on the startup of a subscriber.
-		s.resultCh <- err
-		return
+// HandleEvent updates the handler's state based on register/deregister events.
+func (h *healthServicesHandler) HandleEvent(event *stream.Event) {
+	serviceHealth := event.GetServiceHealth()
```

Comment:
Does this panic or return nil if the event was wrong type?

Reply:
I changed the subscriber to panic if it somehow receives an event of the wrong type

Reply:
How does subscriber know that it's the wrong type? this is the only place that knows which types the health topic should be sending.

I think panic is also dramatic too - it makes any bug with backward compatibility later into a "crash all agents" instead of just" some blocking queries broke. I suggest we error (not just ignore as that leads to silent bad state) the stream but not panic the whole agent.